sw/host/tests/chip/rv_dm/src/access_after_wakeup.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::PathBuf;
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use clap::Parser;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::io::gpio::PinMode;
+use opentitanlib::io::jtag::JtagTap;
+use opentitanlib::test_utils;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::mem::MemWriteReq;
+use opentitanlib::uart::console::UartConsole;
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "600s")]
+    timeout: Duration,
+
+    /// Path to the ELF file being tested on the device.
+    #[arg(long)]
+    firmware_elf: PathBuf,
+}
+
+fn test_access_after_wakeup(
+    opts: &Opts,
+    transport: &TransportWrapper,
+    software_barrier_addr: u32,
+) -> Result<()> {
+    transport.pin_strapping("PINMUX_TAP_RISCV")?.apply()?;
+    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+
+    let power_button = transport.gpio_pin("Ior13")?;
+    power_button.set_mode(PinMode::PushPull)?;
+    power_button.write(true)?;
+
+    // Enable console and wait for the message.
+    let uart = &*transport.uart("console")?;
+    uart.set_flow_control(true)?;
+    let _ = UartConsole::wait_for(uart, r"Running [^\r\n]*", opts.timeout)?;
+    let _ = UartConsole::wait_for(uart, r"Software Setup.", opts.timeout)?;
+
+    // Write to progbuf0 and and confirm readback.
+    let mut jtag = opts
+        .init
+        .jtag_params
+        .create(transport)?
+        .connect(JtagTap::RiscvTap)?
+        .into_raw()?;
+
+    let random_value: u32 = rand::random();
+    log::info!("Writing {random_value:#x} to progbuf0");
+    jtag.execute(&format!("riscv dmi_write 0x20 {random_value:#x}"))?;
+
+    let readback = u32::from_str_radix(
+        jtag.execute("riscv dmi_read 0x20")?
+            .trim()
+            .trim_start_matches("0x"),
+        16,
+    )?;
+    assert_eq!(random_value, readback);
+
+    // Allow the software to continue execution.
+    log::info!("SoftwareBarrier = 1");
+    MemWriteReq::execute(uart, software_barrier_addr, &[1])?;
+
+    // Wait for the software to fall asleep.
+    let _ = UartConsole::wait_for(uart, r"Entering normal sleep", opts.timeout)?;
+
+    std::thread::sleep(Duration::from_secs(1));
```

Comment:
That seems like a very long delay. You could simply wait to see `Entering normal sleep.` on the UART?

Reply:
I was trying to make sure that core actually enters the sleep, but I suppose if I wait for the full message (`"Entering normal sleep.\n"`) then it should be fine.